File: javascript/designpattern/creational/AbstractFactoryPattern.js
/**
 * Abstract Factory Design Pattern
 * 
 * Intent: Provide an interface for creating families of related or dependent objects
 * without specifying their concrete classes.
 * 
 * Use cases:
 * 1. When a system should be independent of how its products are created, composed, and represented
 * 2. When a system should be configured with one of multiple families of products
 * 3. When a family of related product objects is designed to be used together
 * 
 * Components:
 * 1. Abstract Factory
 * 2. Concrete Factories
 * 3. Abstract Products
 * 4. Concrete Products
 */

// Abstract Products (using classes with abstract methods)
class Button {
    render() {
        throw new Error('Method render() must be implemented');
    }

    onClick() {
        throw new Error('Method onClick() must be implemented');
    }
}

class TextBox {
    render() {
        throw new Error('Method render() must be implemented');
    }

    setText(text) {
        throw new Error('Method setText() must be implemented');
    }

    getText() {
        throw new Error('Method getText() must be implemented');
    }
}

class CheckBox {
    render() {
        throw new Error('Method render() must be implemented');
    }

    setChecked(checked) {
        throw new Error('Method setChecked() must be implemented');
    }

    isChecked() {
        throw new Error('Method isChecked() must be implemented');
    }
}

// Concrete Products for Windows
class WindowsButton extends Button {
    render() {
        console.log('Rendering Windows style button');
    }

    onClick() {
        console.log('Windows button clicked');
    }
}

class WindowsTextBox extends TextBox {
    constructor() {
        super();
        this.text = '';
    }

    render() {
        console.log('Rendering Windows style textbox');
    }

    setText(text) {
        this.text = text;
        console.log('Windows textbox text set to:', text);
    }

    getText() {
        return this.text;
    }
}

class WindowsCheckBox extends CheckBox {
    constructor() {
        super();
        this.checked = false;
    }

    render() {
        console.log('Rendering Windows style checkbox');
    }

    setChecked(checked) {
        this.checked = checked;
        console.log('Windows checkbox', checked ? 'checked' : 'unchecked');
    }

    isChecked() {
        return this.checked;
    }
}

// Concrete Products for Mac
class MacButton extends Button {
    render() {
        console.log('Rendering Mac style button');
    }

    onClick() {
        console.log('Mac button clicked');
    }
}

class MacTextBox extends TextBox {
    constructor() {
        super();
        this.text = '';
    }

    render() {
        console.log('Rendering Mac style textbox');
    }

    setText(text) {
        this.text = text;
        console.log('Mac textbox text set to:', text);
    }

    getText() {
        return this.text;
    }
}

class MacCheckBox extends CheckBox {
    constructor() {
        super();
        this.checked = false;
    }

    render() {
        console.log('Rendering Mac style checkbox');
    }

    setChecked(checked) {
        this.checked = checked;
        console.log('Mac checkbox', checked ? 'checked' : 'unchecked');
    }

    isChecked() {
        return this.checked;
    }
}

// Abstract Factory
class GUIFactory {
    createButton() {
        throw new Error('Method createButton() must be implemented');
    }

    createTextBox() {
        throw new Error('Method createTextBox() must be implemented');
    }

    createCheckBox() {
        throw new Error('Method createCheckBox() must be implemented');
    }
}

// Concrete Factories
class WindowsFactory extends GUIFactory {
    createButton() {
        return new WindowsButton();
    }

    createTextBox() {
        return new WindowsTextBox();
    }

    createCheckBox() {
        return new WindowsCheckBox();
    }
}

class MacFactory extends GUIFactory {
    createButton() {
        return new MacButton();
    }

    createTextBox() {
        return new MacTextBox();
    }

    createCheckBox() {
        return new MacCheckBox();
    }
}

// Client Application
class Application {
    constructor(factory) {
        this.button = factory.createButton();
        this.textBox = factory.createTextBox();
        this.checkBox = factory.createCheckBox();
    }

    render() {
        console.log('\nRendering Application UI:');
        this.button.render();
        this.textBox.render();
        this.checkBox.render();
    }

    simulateUserInteraction() {
        console.log('\nSimulating User Interaction:');
        this.button.onClick();
        this.textBox.setText('Hello, World!');
        this.checkBox.setChecked(true);
    }
}

// Alternative implementation using factory functions
const createGUIFactory = (type) => {
    const factories = {
        Windows: () => ({
            createButton: () => new WindowsButton(),
            createTextBox: () => new WindowsTextBox(),
            createCheckBox: () => new WindowsCheckBox()
        }),
        Mac: () => ({
            createButton: () => new MacButton(),
            createTextBox: () => new MacTextBox(),
            createCheckBox: () => new MacCheckBox()
        })
    };

    return factories[type] || (() => {
        throw new Error(`Unsupported UI type: ${type}`);
    });
};

// Test function to demonstrate the pattern
function testAbstractFactoryPattern() {
    console.log('Abstract Factory Pattern Demo\n');

    // Create Windows UI
    console.log('Creating Windows UI Application:');
    const windowsApp = new Application(new WindowsFactory());
    windowsApp.render();
    windowsApp.simulateUserInteraction();

    // Create Mac UI
    console.log('\nCreating Mac UI Application:');
    const macApp = new Application(new MacFactory());
    macApp.render();
    macApp.simulateUserInteraction();

    // Example of using the pattern in a real application
    console.log('\nReal Application Example:');
    console.log('Detecting operating system and creating appropriate UI...');
    
    // Simulate OS detection
    const os = process.platform;
    let factory;
    
    if (os === 'win32') {
        factory = new WindowsFactory();
        console.log('Windows OS detected, creating Windows UI');
    } else if (os === 'darwin') {
        factory = new MacFactory();
        console.log('Mac OS detected, creating Mac UI');
    } else {
        // Default to Windows for other OS
        factory = new WindowsFactory();
        console.log('Unknown OS, defaulting to Windows UI');
    }

    const app = new Application(factory);
    app.render();
    app.simulateUserInteraction();

    // Using the factory function approach
    console.log('\nUsing Factory Function:');
    try {
        const uiType = os === 'darwin' ? 'Mac' : 'Windows';
        const factoryFn = createGUIFactory(uiType)();
        const appFn = new Application(factoryFn);
        appFn.render();
        appFn.simulateUserInteraction();
    } catch (error) {
        console.error(error.message);
    }
}

// Run the tests
testAbstractFactoryPattern(); 

File: java/designpattern/creational/AbstractFactoryPattern.java
/**
 * Abstract Factory Design Pattern
 * 
 * Intent: Provide an interface for creating families of related or dependent objects
 * without specifying their concrete classes.
 * 
 * Use cases:
 * 1. When a system should be independent of how its products are created, composed, and represented
 * 2. When a system should be configured with one of multiple families of products
 * 3. When a family of related product objects is designed to be used together
 * 
 * Components:
 * 1. Abstract Factory
 * 2. Concrete Factories
 * 3. Abstract Products
 * 4. Concrete Products
 */

// Abstract Products
interface Button {
    void render();
    void onClick();
}

interface TextBox {
    void render();
    void setText(String text);
    String getText();
}

interface CheckBox {
    void render();
    void setChecked(boolean checked);
    boolean isChecked();
}

// Concrete Products for Windows
class WindowsButton implements Button {
    @Override
    public void render() {
        System.out.println("Rendering Windows style button");
    }

    @Override
    public void onClick() {
        System.out.println("Windows button clicked");
    }
}

class WindowsTextBox implements TextBox {
    private String text = "";

    @Override
    public void render() {
        System.out.println("Rendering Windows style textbox");
    }

    @Override
    public void setText(String text) {
        this.text = text;
        System.out.println("Windows textbox text set to: " + text);
    }

    @Override
    public String getText() {
        return text;
    }
}

class WindowsCheckBox implements CheckBox {
    private boolean checked = false;

    @Override
    public void render() {
        System.out.println("Rendering Windows style checkbox");
    }

    @Override
    public void setChecked(boolean checked) {
        this.checked = checked;
        System.out.println("Windows checkbox " + (checked ? "checked" : "unchecked"));
    }

    @Override
    public boolean isChecked() {
        return checked;
    }
}

// Concrete Products for Mac
class MacButton implements Button {
    @Override
    public void render() {
        System.out.println("Rendering Mac style button");
    }

    @Override
    public void onClick() {
        System.out.println("Mac button clicked");
    }
}

class MacTextBox implements TextBox {
    private String text = "";

    @Override
    public void render() {
        System.out.println("Rendering Mac style textbox");
    }

    @Override
    public void setText(String text) {
        this.text = text;
        System.out.println("Mac textbox text set to: " + text);
    }

    @Override
    public String getText() {
        return text;
    }
}

class MacCheckBox implements CheckBox {
    private boolean checked = false;

    @Override
    public void render() {
        System.out.println("Rendering Mac style checkbox");
    }

    @Override
    public void setChecked(boolean checked) {
        this.checked = checked;
        System.out.println("Mac checkbox " + (checked ? "checked" : "unchecked"));
    }

    @Override
    public boolean isChecked() {
        return checked;
    }
}

// Abstract Factory
interface GUIFactory {
    Button createButton();
    TextBox createTextBox();
    CheckBox createCheckBox();
}

// Concrete Factories
class WindowsFactory implements GUIFactory {
    @Override
    public Button createButton() {
        return new WindowsButton();
    }

    @Override
    public TextBox createTextBox() {
        return new WindowsTextBox();
    }

    @Override
    public CheckBox createCheckBox() {
        return new WindowsCheckBox();
    }
}

class MacFactory implements GUIFactory {
    @Override
    public Button createButton() {
        return new MacButton();
    }

    @Override
    public TextBox createTextBox() {
        return new MacTextBox();
    }

    @Override
    public CheckBox createCheckBox() {
        return new MacCheckBox();
    }
}

// Client Application
class Application {
    private Button button;
    private TextBox textBox;
    private CheckBox checkBox;

    public Application(GUIFactory factory) {
        button = factory.createButton();
        textBox = factory.createTextBox();
        checkBox = factory.createCheckBox();
    }

    public void render() {
        System.out.println("\nRendering Application UI:");
        button.render();
        textBox.render();
        checkBox.render();
    }

    public void simulateUserInteraction() {
        System.out.println("\nSimulating User Interaction:");
        button.onClick();
        textBox.setText("Hello, World!");
        checkBox.setChecked(true);
    }
}

public class AbstractFactoryPattern {
    public static void main(String[] args) {
        System.out.println("Abstract Factory Pattern Demo\n");

        // Create Windows UI
        System.out.println("Creating Windows UI Application:");
        Application windowsApp = new Application(new WindowsFactory());
        windowsApp.render();
        windowsApp.simulateUserInteraction();

        // Create Mac UI
        System.out.println("\nCreating Mac UI Application:");
        Application macApp = new Application(new MacFactory());
        macApp.render();
        macApp.simulateUserInteraction();

        // Example of using the pattern in a real application
        System.out.println("\nReal Application Example:");
        System.out.println("Detecting operating system and creating appropriate UI...");
        
        // Simulate OS detection
        String os = System.getProperty("os.name").toLowerCase();
        GUIFactory factory;
        
        if (os.contains("windows")) {
            factory = new WindowsFactory();
            System.out.println("Windows OS detected, creating Windows UI");
        } else if (os.contains("mac")) {
            factory = new MacFactory();
            System.out.println("Mac OS detected, creating Mac UI");
        } else {
            // Default to Windows for other OS
            factory = new WindowsFactory();
            System.out.println("Unknown OS, defaulting to Windows UI");
        }

        Application app = new Application(factory);
        app.render();
        app.simulateUserInteraction();
    }
} 